source/extensions/filters/network/http_connection_manager/config.cc

```diff
@@ -524,7 +524,7 @@ void HttpConnectionManagerConfig::processFilter(
       proto_config, context_.messageValidationVisitor(), factory);
   Http::FilterFactoryCb callback =
       factory.createFilterFactoryFromProto(*message, stats_prefix_, context_);
-  bool is_terminal = factory.isTerminalFilter();
+  bool is_terminal = factory.isTerminalFilterByProto(*message, context_);
```

Comment:
Here we should read the message before validating the filter whether it's a terminal filter.